Fix example manifests for webhooks-based scaling (#1354)
actions-runner-controller/actions-runner-controller: README.md

```console
$ helm upgrade --install --namespace actions-runner-system --create-namespace \
             --wait actions-runner-controller actions-runner-controller/actions-runner-controller \

>              --set "githubWebhookServer.enabled=true"

Reviewer: ```suggestion
             --set "githubWebhookServer.enabled=true,service.type=NodePort,githubWebhookServer.ports[0].nodePort=33080"
```

I prefer leaving NodePort as a first option introduced this document, so that K8s beginners can hopefully see you usually need to expose the nodeport anyway, either for the ingress controller or ARC, to let GitHub successfully send webhooks to your webhook endpoint.

Author: There is only one thing that is puzzling me about this, I was under the impression that GitHub will try to use SSL, and the "standard" way of setting up SSL on k8s is with an ingress. 

By exposing a node port the expectation is that the user will have to set up SSL on its own using something outside k8s?

That said I have no problem applying this change, but I am still curious.

Author: Change applied, but I am still curious =P

Reviewer: > By exposing a node port the expectation is that the user will have to set up SSL on its own using something outside k8s?

@VinGarcia Mostly! I tend to provision an external load balancer outside of the K8s cluster, so that it can provide us a stable hostname and TLS termination. I'm usually an AWS shop so I often use AWS ALB for that. NodePort is used to provide access from the load balancer to nodes. In that case you don't even need to setup an ingress controller.

In many cases, it's almost about which component you'd like to expose via NodePort. If you use e.g. ingress-nginx as an ingress controller, you'd need to expose the ingress-controller pods via a NodePort service first, so that your load balancer can finally expose it to the internet(or set `type: LoadBalancer` so that it becomes a Service LoadBalancer, but under the hood it still use NodePort and the difference is that the k8s cloud provider creates an external loadbalancer for you...).

Also note that there are some advanced CNI plugins like https://github.com/aws/amazon-vpc-cni-k8s which are capable of connecting pods, nodes, and load balancers in a flat network. With that, you don't even need NodePort or Ingress. Your external loadbalancer can directly access your pods.

Author: I see, thanks for the detailed explanation @mumoshu =]. I am still learning a lot about k8s, so thanks for sharing.

```

> The command above will create a new deployment and a service for receiving Github Webhooks on the `actions-runner-system` namespace.

Reviewer: @VinGarcia @toast-gear I think this might be a more sustainable way to introduce this guide. We'd like to make it helpful enough to cover more setups, while preventing the maintainers from being overloaded with a lot of questions unrelated to ARC itself.
WDYT?

> The command above will create a new deployment and a service for receiving Github Webhooks on the `actions-runner-system` namespace.

Author: Uhm, I am not against it, but I would phrase it differently because in this way it implies that there is some unspecified problem that should be expected if the user decides to use an ingress.

Maybe I would only change the CAUTION section to something like:

```
> **CAUTION:** The Kubernetes ingress controllers describe below is just a suggestion from the community and
> the ARC team will not provide any user support for ingress controllers as it's not a part of this project.
>
> The following guide has been contributed by the awesome ARC community and provided here as-is. You may be able to ask the community in GitHub Discussions though.
```

> The command above will create a new deployment and a service for receiving Github Webhooks on the `actions-runner-system` namespace.

Reviewer: > there is some unspecified problem that should be expected if the user decides to use an ingress

Actually, that's my belief. Using already set up ingress controller is easy but setting it up isn't actually easy as you usually need to know how to expose the ingress controller itself, and to do so know how your pod and node network work and how the loadbalancer redirects or proxies the packet to the ingress controller..

> The command above will create a new deployment and a service for receiving Github Webhooks on the `actions-runner-system` namespace.

Reviewer: So, if you're also a cluster-admin, service loadbalancers or nodeports are simpler as there are less moving parts than setting up and using ingress controllers. That's why I preferred suggesting the use of NodePort in the doc earlier than Ingress. WDYT?

> The command above will create a new deployment and a service for receiving Github Webhooks on the `actions-runner-system` namespace.

Author: I see, so just reiterating my other comment, this change makes a lot of sense to me. So I am in favor of it.

I just suggested changing the `**CAUTION**` section because I thought it would give an impression that using the ingress was necessarily a wrong approach. But I would keep all the rest of it exactly as you wrote it.

From this point on I don't have a strong opinion in regards to keeping it exactly as you wrote or making the change on the `CAUTION` part, so just say the word about which one you prefer and I will make the change, I believe this is an important addition regardless of which version we decide upon.

> The command above will create a new deployment and a service for receiving Github Webhooks on the `actions-runner-system` namespace.

Reviewer: > I just suggested changing the **CAUTION** section because I thought it would give an impression that using the ingress was necessarily a wrong approach. But I would keep all the rest of it exactly as you wrote it.

Thanks- I see! I started to feel like it's also aligned with my intention. The "use at your own risk" message might be too cautious. Let's merge my suggestion and your suggestion.

> The command above will create a new deployment and a service for receiving Github Webhooks on the `actions-runner-system` namespace.

Author: Great, I just merged them and pushed the changes =]

spec:
  tls:
  - hosts:
    - your.domain.com
    secretName: your-tls-secret-name
  rules:
    - http:
        paths:
          - path: /actions-runner-controller-github-webhook-server
            pathType: Prefix
            backend:
              service:
                name: actions-runner-controller-github-webhook-server
                port:
                  number: 80
```

Make sure to set the `spec.tls.secretName` to the name of your TLS secret and
`spec.tls.hosts[0]` to your own domain.

Then create this resource on your cluster with the following command:

```bash
kubectl apply -n actions-runner-system -f arc-webhook-server.yaml
```

After this step your webhook server should be ready to start receiving webhooks from Github.

To configure Github to start sending you webhooks, go to the settings page of your repository
or organization then click on `Webhooks`, then on `Add webhook`.

There set the "Payload URL" field with the webhook URL you just created,
if you followed the example ingress above the URL would be something like this:

- https://your.domain.com/actions-runner-controller-github-webhook-server

> Remember to replace `your.domain.com` with your own domain.

> Then click on "let me select individual events" and choose:

Reviewer: ```suggestion
Then click on "let me select individual events" and choose `Workflow Jobs`.

You may also want to choose the following event(s) if you use it as a scale trigger in your HRA spec:
```

> Then click on "let me select individual events" and choose:
> 
> - Check runs
> - Pushes
> - Pull Requests
> - Workflow Jobs

Reviewer: ```suggestion
```suggestion
Then click on "let me select individual events" and choose `Workflow Jobs`.

You may also want to choose the following event(s) if you use it as a scale trigger in your HRA spec:

- Check runs
- Pushes
- Pull Requests
```

We recommend using `Workflow Jobs` only these days so better introduce it as a standard way :)

Author: Applied!

Later you can remove any of these you are not using to reduce the amount of data sent to your server.

Then click on `Add Webhook`.

> Github will then send a `ping` event to your webhook server to check if it is working, if it is you'll see a green V mark

Author: Applied! I also did a quick search & replace so all instances of `Github` were replaced for `GitHub`

This webhook should cover most people's needs, please experiment with this webhook first before considering the others.

```yaml
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: example-runners

>   namespace: actions-runner-system

Reviewer: ```suggestion
```

I believe it's a good practice to separate the controller and runner namespace for security.
Just keep omitting it so that it would default to the `default` namespace

Author: Uhm, I decided to test whether this really was unnecessary, but when I removed it I started having some problems. It's late now, so I might have made some mistakes, I will continue this investigation tomorrow.

Author: @mumoshu I am not sure how to interpret the results of my experiment, but I tried removing the namespace from the `RunnerDeployment` and the `HorizontalRunnerAutoscaler` and it actually stopped working.

Just so you can understand the steps I took, I used this command to install arc:

```bash
helm upgrade --install --namespace actions-runner-system --create-namespace \
             --wait actions-runner-controller actions-runner-controller/actions-runner-controller \
             --set "githubWebhookServer.enabled=true,githubWebhookServer.ports[0].nodePort=33080"
```

Then I created these two resources below, originally with the namespace, then I removed them then I added them again.

```diff
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: example-runners
-  namespace: actions-runner-system
spec:
  template:
    spec:
      repository: vingarcia/ksql
---
apiVersion: actions.summerwind.dev/v1alpha1
kind: HorizontalRunnerAutoscaler
metadata:
  name: example-runners
-  namespace: actions-runner-system
spec:
  scaleDownDelaySecondsAfterScaleOut: 300
  minReplicas: 1
  maxReplicas: 10
  scaleTargetRef:
    name: example-runners
  scaleUpTriggers:
  - githubEvent: {}
    duration: "30m"
```

After removing by running the delete kubectl command on both resources I was left with no RunnerDeployments and no HRAs, then I created them again without the namespace and provoked some workflowJob events, and it did not autoscale as it was autoscaling before. I checked and both resources were created in the `default` namespace then I opened the logs for the `actions-runner-controller` and the `actions-runner-controller-github-webhook-server` and provoked another workflowJob event (by pushing some changes to my repo). On the controller I got nothing out of the usual, but I got several errors on the webhook-server:

```logs
2022-05-17T23:21:06Z    ERROR   controllers.webhookbasedautoscaler      finding repository-wide runner  {"event": "workflow_job", "hookID": "353945701", "delivery": "069f0440-d638-11ec-8ec3-d741f7b474a4", "workflowJob.status": "queued", "workflowJob.labels": ["self-hosted"], "repository.name": "ksql", "repository.owner.login": "VinGarcia", "repository.owner.type": "User", "enterprise.slug": "", "action": "queued", "repository": "VinGarcia/ksql", "error": "cache contained <nil>, which is not an Object"}
github.com/actions-runner-controller/actions-runner-controller/controllers.(*HorizontalRunnerAutoscalerGitHubWebhook).getJobScaleUpTargetForRepoOrOrg
        /workspace/controllers/horizontal_runner_autoscaler_webhook.go:456
github.com/actions-runner-controller/actions-runner-controller/controllers.(*HorizontalRunnerAutoscalerGitHubWebhook).Handle
        /workspace/controllers/horizontal_runner_autoscaler_webhook.go:236
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:2047
net/http.(*ServeMux).ServeHTTP
        /usr/local/go/src/net/http/server.go:2425
net/http.serverHandler.ServeHTTP
        /usr/local/go/src/net/http/server.go:2879
net/http.(*conn).serve
        /usr/local/go/src/net/http/server.go:1930
2022-05-17T23:21:06Z    ERROR   controllers.webhookbasedautoscaler      handling check_run event        {"event": "workflow_job", "hookID": "353945701", "delivery": "069f0440-d638-11ec-8ec3-d741f7b474a4", "workflowJob.status": "queued", "workflowJob.labels": ["self-hosted"], "repository.name": "ksql", "repository.owner.login": "VinGarcia", "repository.owner.type": "User", "enterprise.slug": "", "action": "queued", "error": "cache contained <nil>, which is not an Object"}
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:2047
net/http.(*ServeMux).ServeHTTP
        /usr/local/go/src/net/http/server.go:2425
net/http.serverHandler.ServeHTTP
        /usr/local/go/src/net/http/server.go:2879
net/http.(*conn).serve
        /usr/local/go/src/net/http/server.go:1930
```

Then I copied the logs, deleted both resources and created them again this time with the namespace field set, and it worked normally, the logs showed normal event processing messages:

```logs
2022-05-17T23:30:46Z    DEBUG   controllers.webhookbasedautoscaler      Found 1 HRAs by key     {"key": "VinGarcia/ksql"}
2022-05-17T23:30:46Z    INFO    controllers.webhookbasedautoscaler      job scale up target is repository-wide runners  {"event": "workflow_job", "hookID": "353945701", "delivery": "60751ad0-d639-11ec-823d-b0f433c135ab", "workflowJob.status": "queued", "workflowJob.labels": ["self-hosted"], "repository.name": "ksql", "repository.owner.login": "VinGarcia", "repository.owner.type": "User", "enterprise.slug": "", "action": "queued", "repository": "ksql"}
2022-05-17T23:30:46Z    DEBUG   controllers.webhookbasedautoscaler      Patching hra example-runners for capacityReservations update    {"before": 0, "expired": 0, "amount": 1, "after": 1}
2022-05-17T23:30:46Z    INFO    controllers.webhookbasedautoscaler      scaled example-runners by 1
```

Reviewer: @VinGarcia Thanks for your detailed report 🙏 

The error you encountered, `cache contained <nil>, which is not an Object`, is very interesting. It reminds me of the same looking error that we concluded to occur due to possibly a hidden bug in `controller-runtime` which we use as the framework for K8s controllers(like ARC).

To see the nature of the problem, would you mind recreating the runnerdeployments "without namespace" once again, and restart ARC, by removing the controller-manager pod and webhookbasedautoscaler 's pod manually? (THose pods are managed by K8s Deployments so K8s would automatically bring up another pods as replacements, hence restart)

Author: Oh sure, so, I just made this experiment, and it worked this time (although it took some time to actually start working, at first it was reporting `Found 0 HRAs by key     {"key": "VinGarcia/ksql"}`, and only after some time it started detecting the HRA and scaling normally.

It was a good thing that I waited for a while before writing down the results.

Ok, so I think we can remove the namespace too.

spec:
  template:
    spec:

>       labels:
>         - self-hosted

Reviewer: ```suggestion
```

AFAIK, you don't need to explicitly set the `self-hosted` label as it's implied and forced by `actions/runner`/GitHub

Author: I think you are right, I was pretty sure it was not working before I added this line, but I just tested without it and it worked just as well.

      repository: example/myrepo
---
apiVersion: actions.summerwind.dev/v1alpha1
kind: HorizontalRunnerAutoscaler
metadata:
  name: example-runners

>   namespace: actions-runner-system

Reviewer: ```suggestion
```

ref https://github.com/actions-runner-controller/actions-runner-controller/pull/1354/files#r873273924

          value: "true"
        # Configure runner with legacy --once instead of --ephemeral flag
        # WARNING | THIS ENV VAR IS DEPRECATED AND WILL BE REMOVED

>         # IN A FUTURE VERSION OF ARC. SEE ISSUE #1196 FOR DETAILS

Reviewer: ```suggestion
        # THIS ENV VAR WILL BE REMOVED SOON.
        # SEE ISSUE #1196 FOR DETAILS
```
